asserts/cluster.go

```diff
@@ -0,0 +1,323 @@
+
+		result = append(result, d)
+	}
+	return result, nil
+}
+
+func checkClusterSnap(snap map[string]any) (ClusterSnap, error) {
+	state, err := checkNotEmptyString(snap, "state")
+	if err != nil {
+		return ClusterSnap{}, err
+	}
+
+	if !strutil.ListContains(validClusterSnapStates, state) {
+		return ClusterSnap{}, fmt.Errorf("snap state must be one of %v", validClusterSnapStates)
+	}
+
+	instance, err := checkNotEmptyString(snap, "instance")
+	if err != nil {
+		return ClusterSnap{}, err
+	}
+
+	if err := naming.ValidateInstance(instance); err != nil {
+		return ClusterSnap{}, fmt.Errorf("invalid snap instance name: %v", err)
+	}
+
+	ch, err := checkNotEmptyString(snap, "channel")
+	if err != nil {
+		return ClusterSnap{}, err
+	}
+
+	if _, err := channel.Parse(ch, ""); err != nil {
+		return ClusterSnap{}, fmt.Errorf("invalid channel name %q: %v", ch, err)
+	}
+
+	return ClusterSnap{
+		State:    state,
+		Instance: instance,
+		Channel:  ch,
+	}, nil
+}
+
+func checkClusterSnaps(snaps []any) ([]ClusterSnap, error) {
+	result := make([]ClusterSnap, 0, len(snaps))
+	for _, entry := range snaps {
+		snap, ok := entry.(map[string]any)
+		if !ok {
+			return nil, errors.New(`"snaps" field must be a list of maps`)
+		}
+
+		s, err := checkClusterSnap(snap)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, s)
+	}
+
+	return result, nil
+}
+
+func checkClusterSubcluster(subcluster map[string]any) (ClusterSubcluster, error) {
+	name, err := checkNotEmptyString(subcluster, "name")
+	if err != nil {
+		return ClusterSubcluster{}, err
+	}
+
+	devices, err := checkStringList(subcluster, "devices")
+	if err != nil {
+		return ClusterSubcluster{}, err
+	}
+
+	ids := make([]int, 0, len(devices))
+	for _, dev := range devices {
+		id, err := atoi(dev, "device id %q", dev)
+		if err != nil {
+			return ClusterSubcluster{}, err
+		}
+		ids = append(ids, id)
+	}
+
+	list, err := checkList(subcluster, "snaps")
+	if err != nil {
+		return ClusterSubcluster{}, err
+	}
+
+	snaps, err := checkClusterSnaps(list)
+	if err != nil {
+		return ClusterSubcluster{}, err
+	}
+
+	return ClusterSubcluster{
+		Name:    name,
+		Devices: ids,
+		Snaps:   snaps,
+	}, nil
+}
+
+func checkClusterSubclusters(subclusters []any) ([]ClusterSubcluster, error) {
+	result := make([]ClusterSubcluster, 0, len(subclusters))
+	for _, entry := range subclusters {
+		subcluster, ok := entry.(map[string]any)
+		if !ok {
+			return nil, errors.New(`"subclusters" field must be a list of maps`)
+		}
+
+		s, err := checkClusterSubcluster(subcluster)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, s)
+	}
+
+	return result, nil
+}
+
+func validateClusterDeviceIDs(devices []ClusterDevice, subclusters []ClusterSubcluster) error {
+	seen := make(map[int]bool, len(devices))
+	for _, device := range devices {
+		seen[device.ID] = true
+	}
+
+	for _, subcluster := range subclusters {
+		for _, id := range subcluster.Devices {
+			if seen[id] {
```

Comment:
Missing a negation in this condition

```suggestion
			if !seen[id] {
```

Reply:
Ah whoops, thanks. I swapped it out from a `map[int]struct{}` and messed up this bit.